src/main/java/org/biouno/unochoice/util/SafeHtmlExtendedMarkupFormatter.java

```diff
@@ -0,0 +1,47 @@
+package org.biouno.unochoice.util;
+
+import com.google.common.base.Throwables;
+import hudson.markup.BasicPolicy;
+import hudson.markup.MarkupFormatter;
+import org.owasp.html.HtmlPolicyBuilder;
+import org.owasp.html.HtmlSanitizer;
+import org.owasp.html.HtmlStreamRenderer;
+import org.owasp.html.PolicyFactory;
+
+import java.io.IOException;
+import java.io.Writer;
+
+public class SafeHtmlExtendedMarkupFormatter extends MarkupFormatter {
+
+    public static final SafeHtmlExtendedMarkupFormatter INSTANCE = new SafeHtmlExtendedMarkupFormatter();
+
+    /**
+     * {@link BasicPolicy#POLICY_DEFINITION} is the policy used by {@link hudson.markup.RawHtmlMarkupFormatter}.
+     * We start from that secure policy and then extend it to include required elements for this plugin.
+     */
+    private static final PolicyFactory POLICY = BasicPolicy.POLICY_DEFINITION.and(new HtmlPolicyBuilder()
+        .allowElements("input", "textarea", "select", "option")
+        .allowAttributes("id", "class", "style", "type", "name", "value", "placeholder", "disabled", "checked", "max", "maxlength", "min", "minlength", "multiple", "pattern", "readonly", "step").onElements("input")
+        .allowAttributes("id", "class", "style", "maxlength", "name", "placeholder", "disabled", "readonly", "wrap", "rows", "cols").onElements("textarea")
+        .allowAttributes("id", "class", "style", "disabled", "multiple", "name", "required", "size").onElements("select")
+        .allowAttributes("id", "class", "style", "disabled", "label", "selected", "value").onElements("option")
```

Comment:
FWIW one of the motivations of antisamy-markup-formatter 2.0's new restrictions was to prevent regular users from creating UI that looks like native Jenkins UI. This could allow regular users to trick admins into performing actions they never intended.

This code looks like it would allow quite a bit of that. Ideally someone would investigate this in more detail (I wonder whether the problem is re-introduced to the same extent or whether `form` element not being allowed sidesteps it completely), but failing that I would recommend it be documented that this plugin allows doing just that. It could be advertised as a feature even, plus a note about the potential for mischief :-)

Reply:
That was my concern @daniel-beck . The idea of the formatter parameter type is exactly to display either UI that looks like Jenkins UI, or custom elements like an image or text. So that's kind of a feature I think.

Nevertheless, if an admin doesn't use it with care it could indeed be misused. So will add to the top of the documentation a note, and an entry in the 2.4 release notes, drop a note to a mailing list with some users (biouno-users), plus include in the tweet message to users.

Reply:
By 'formatter parameter type'. Do you mean (reactive) 'Reference' type @kinow ?